File: src/main/scala/dev/helium_build/record/ArchiveRecorder.scala
package dev.helium_build.record

import java.io.{File, FileInputStream, FileOutputStream}
import java.nio.charset.StandardCharsets
import java.nio.file.Files

import cats.effect.Resource
import dev.helium_build.sdk.{SdkInfo, SdkInstallManager, SdkLoader}
import io.circe.{Json, JsonObject}
import org.apache.commons.compress.archivers.tar.{TarArchiveEntry, TarArchiveOutputStream, TarConstants}
import org.apache.commons.io.{FileUtils, IOUtils}
import zio.blocking.Blocking
import zio.{IO, Managed, RIO, Ref, Semaphore, Task, ZIO, ZManaged}
import zio.interop.catz._
import cats.implicits._
import dev.helium_build.build.BuildSchema
import dev.helium_build.conf.RepoConfig
import dev.helium_build.util.ArchiveUtil

final class ArchiveRecorder[R <: Blocking] private
(
  lock: Semaphore,
  recordedSchema: Ref[Option[String]],
  recordedRepoConfig: Ref[Option[String]],
  recordedSdks: Ref[Set[String]],
  recordedArtifacts: Ref[Set[String]],
  recordedMetadata: Ref[Map[String, Json]],
  archive: TarArchiveOutputStream,
  cacheDir: File,
  protected override val sdkDir: File,
  protected override val schemaFile: File,
  override val workDir: File,
  protected override val confDir: File,
) extends LiveRecorder[R] {
  import ArchiveRecorder._

  private def writeTextFile(path: String, content: String): RIO[R, Unit] =
    ZIO.accessM[R] { _.blocking.effectBlocking {
      val entry = new TarArchiveEntry(path)
      val bytes = content.getBytes(StandardCharsets.UTF_8)
      entry.setSize(bytes.length)
      archive.putArchiveEntry(entry)

      IOUtils.write(bytes, archive)
      archive.closeArchiveEntry()
    } }

  override protected def cacheBuildSchema(readBuildSchema: RIO[R, String]): RIO[R, String] =
    lock.withPermit(
      recordedSchema.get.flatMap {
        case Some(schema) => IO.succeed(schema)
        case None =>
          for {
            data <- readBuildSchema
            _ <- recordedSchema.set(Some(data))
            _ <- writeTextFile(buildSchemaPath, data)
          } yield data
      }
    )

  override def sdkInstallManager: ZIO[R, Throwable, SdkInstallManager] = for {
    sdkInstallManager <- SdkInstallManager(cacheDir)
  } yield new SdkInstallManager {
    override def getInstalledSdkDir(sdk: SdkInfo): RIO[Blocking, (String, File)] = lock.withPermit {
      val sdkHash = SdkLoader.sdkSha256(sdk)
      recordedSdks.get.flatMap { sdks =>
        if(sdks.contains(sdkHash)) {
          sdkInstallManager.getInstalledSdkDir(sdk)
        } else {
          for {
            (hash, dir) <- sdkInstallManager.getInstalledSdkDir(sdk)
            _ <- ArchiveUtil.addDirToArchive(archive, sdkPath(hash), dir.getParentFile)
            _ <- recordedSdks.set(sdks + hash)
          } yield (hash, dir)
        }
      }
    }
  }


  override protected def cacheRepoConfig(readRepoConfig: RIO[R, String]): RIO[R, String] =
    lock.withPermit(
      recordedRepoConfig.get.flatMap {
        case Some(config) => IO.succeed(config)
        case None =>
          for {
            data <- readRepoConfig
            _ <- recordedRepoConfig.set(Some(data))
            _ <- writeTextFile(repoConfigPath, data)
          } yield data
      }
    )

  override def recordTransientMetadata(path: String)(fetch: ZIO[R, Throwable, Json]): ZIO[R, Throwable, Json] = {
    lock.withPermit(
      recordedMetadata.get.flatMap { metadata =>
        metadata.get(path) match {
          case Some(value) =>
            IO.succeed(value)
          case None =>
            for {
              value <- fetch
              _ <- recordedMetadata.set(metadata + (path -> value))
            } yield value
        }
      }
    )
  }


  def recordArtifact(path: String)(fetch: File => ZIO[R, Throwable, File]): ZIO[R, Throwable, File] =
    lock.withPermit(
      recordedArtifacts.get.flatMap { artifacts =>
        if(artifacts.contains(path))
          fetch(cacheDir)
        else
          for {
            file <- fetch(cacheDir)
            _ <- ArchiveUtil.addFileToArchive(archive, artifactPath(path), file)
            _ <- recordedArtifacts.set(artifacts + path)
          } yield file
      }
    )


  private def writeMetadata: ZIO[R, Throwable, Unit] =
    recordedMetadata.get.flatMap { metadata =>
      ZIO.accessM { _.blocking.effectBlocking {
        val entry = new TarArchiveEntry(transientMetadataPath)
        val data = Json.fromJsonObject(JsonObject.fromMap(metadata)).noSpaces.getBytes(StandardCharsets.UTF_8)
        entry.setSize(data.length)
        archive.putArchiveEntry(entry)
        IOUtils.write(data, archive)
        archive.closeArchiveEntry()
      } }
    }


}

object ArchiveRecorder {

  def apply[R <: Blocking]
  (
    cacheDir: File,
    sdkDir: File,
    schemaFile: File,
    archiveFile: File,
    workDir: File,
    confDir: File,
  ): ZManaged[R, Throwable, ArchiveRecorder[R]] =
    for {
      outStream <- ZManaged.fromAutoCloseable(ZIO.accessM[R] { _.blocking.effectBlocking { new FileOutputStream(archiveFile) }})
      tarStream <- ZManaged.fromAutoCloseable(ZIO.accessM[R] { _.blocking.effectBlocking { new TarArchiveOutputStream(outStream) }})
      _ <- ZManaged.fromEffect(IO.effect { tarStream.setLongFileMode(TarArchiveOutputStream.LONGFILE_POSIX) })
      _ <- ZManaged.fromEffect(ArchiveUtil.addDirToArchive(tarStream, workDirPath, workDir))
      recorder <- ZManaged.make[R, Throwable, ArchiveRecorder[R]](
        for {
          lock <- Semaphore.make(1)
          recordedSchema <- Ref.make(Option.empty[String])
          recordedRepoConfig <- Ref.make(Option.empty[String])
          recordedSdks <- Ref.make(Set.empty[String])
          recordedArtifacts <- Ref.make(Set.empty[String])
          recordedMetadata <- Ref.make(Map.empty[String, Json])
        } yield new ArchiveRecorder[R](
          lock,
          recordedSchema,
          recordedRepoConfig,
          recordedSdks,
          recordedArtifacts,
          recordedMetadata,
          tarStream,
          cacheDir = cacheDir,
          sdkDir = sdkDir,
          schemaFile = schemaFile,
          workDir = workDir,
          confDir = confDir,
        )
      )(_.writeMetadata.orDie)
    } yield recorder


  val buildSchemaPath = "build.toml"
  val repoConfigPath = "conf/repos.toml"
  val transientMetadataPath = "dependencies-metadata.json"
  val workDirPath = "work"
  def sdkPath(hash: String) = "sdks/" + hash
  def artifactPath(path: String) = "dependencies/" + path


}

File: src/Engine/Record/ArchiveRecorder.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using Helium.Engine.Cache;
using Helium.Sdks;
using Helium.Util;
using ICSharpCode.SharpZipLib.Tar;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Nito.AsyncEx;

namespace Helium.Engine.Record
{
    internal class ArchiveRecorder : LiveRecorder
    {
        private ArchiveRecorder(string cacheDir, string sdkDir, string schemaFile, string sourcesDir, string confDir, Stream tarDataStream, TarOutputStream tarStream) {
            this.cacheDir = cacheDir;
            this.tarDataStream = tarDataStream;
            this.tarStream = tarStream;
            SdkDir = sdkDir;
            SchemaFile = schemaFile;
            SourcesDir = sourcesDir;
            ConfDir = confDir;
        }

        private readonly string cacheDir;
        private readonly Stream tarDataStream;
        private readonly ConcurrentDictionary<string, Task<JObject>> metadataCache = new ConcurrentDictionary<string, Task<JObject>>();
        
        private readonly AsyncLock tarLock = new AsyncLock();
        private readonly TarOutputStream tarStream;
        private readonly HashSet<string> recordedSdks = new HashSet<string>();
        private readonly HashSet<string> recordedArtifacts = new HashSet<string>();
        private bool hasRecordedSchema;
        private bool hasRecordedRepoConfig;

        public override string SourcesDir { get; }
        protected override string SchemaFile { get; }
        protected override string SdkDir { get; }
        protected override string ConfDir { get; }


        public override async ValueTask DisposeAsync() {
            try {
                await tarStream.DisposeAsync();
            }
            finally {
                await tarDataStream.DisposeAsync();
            }
        }

        public override ISdkInstallManager CreateSdkInstaller() =>
            new ArchiveSdkInstallManager(this);

        public override async Task<string> RecordArtifact(string path, Func<string, Task<string>> fetch) {
            var file = await fetch(cacheDir);

            using var _ = await tarLock.LockAsync();

            if(!recordedArtifacts.Contains(path)) {
                await ArchiveUtil.AddFileToTar(tarStream, ArtifactPath(path), file);
                recordedArtifacts.Add(path);
            }
            
            return file;
        }


        public override Task<JObject> RecordTransientMetadata(string path, Func<Task<JObject>> fetch) =>
            metadataCache.GetOrAdd(path, _ => fetch());


        protected override async Task<string> RecordBuildSchema(Func<Task<string>> readBuildSchema) {
            var result = await readBuildSchema();

            using var _ = await tarLock.LockAsync();
            if(!hasRecordedSchema) {
                await ArchiveUtil.AddFileToTar(tarStream, BuildSchemaPath, SchemaFile);
                hasRecordedSchema = true;
            }
            
            return result;
        }

        protected override async Task<string> RecordRepoConfig(Func<Task<string>> readRepoConfig) {
            var result = await readRepoConfig();

            using var _ = await tarLock.LockAsync();
            if(!hasRecordedRepoConfig) {
                await ArchiveUtil.AddFileToTar(tarStream, RepoConfigPath, Path.Combine(ConfDir, "repos.toml"));
                hasRecordedRepoConfig = true;
            }
            
            return result;
        }

        public override async Task RecordMetadata() {
            var obj = new JObject();
            foreach(var (key, value) in metadataCache) {
                obj[key] = await value;
            }

            await ArchiveUtil.AddStringToTar(tarStream, TransientMetadataPath, JsonConvert.SerializeObject(obj));
        }

        private sealed class ArchiveSdkInstallManager : SdkInstallManager
        {
            public ArchiveSdkInstallManager(ArchiveRecorder archiveRecorder) : base(archiveRecorder.cacheDir) {
                this.archiveRecorder = archiveRecorder;
            }

            private readonly ArchiveRecorder archiveRecorder;
            
            public override async Task<(string hash, string installDir)> GetInstalledSdkDir(SdkInfo sdk) {
                var (hash, installDir) = await base.GetInstalledSdkDir(sdk);

                using var _ = await archiveRecorder.tarLock.LockAsync();
                
                if(!archiveRecorder.recordedSdks.Contains(hash)) {
                    await ArchiveUtil.AddDirToTar(archiveRecorder.tarStream, SdkPath(hash), Path.GetFullPath(Path.Combine(installDir, "..")));
                    archiveRecorder.recordedSdks.Add(hash);
                }
                
                return (hash, installDir);
            }
        }

        public static async Task<IRecorder> Create(string cacheDir, string sdkDir, string schemaFile, string sourcesDir, string confDir, string archiveFile) {
            var stream = File.Create(archiveFile);
            TarOutputStream tarStream;
            try {
                tarStream = new TarOutputStream(stream);
                try {
                    await ArchiveUtil.AddDirToTar(tarStream, SourcesPath, sourcesDir);
                }
                catch {
                    try {
                        await tarStream.DisposeAsync();
                    }
                    catch {}
                    throw;
                }
            }
            catch {
                try {
                    await stream.DisposeAsync();
                }
                catch {}
                throw;
            }
            
            return new ArchiveRecorder(
                cacheDir: cacheDir,
                sdkDir: sdkDir,
                schemaFile: schemaFile,
                sourcesDir: sourcesDir,
                confDir: confDir,
                tarDataStream: stream,
                tarStream: tarStream
            );
        }

        public const string BuildSchemaPath = "build.toml";
        public const string RepoConfigPath = "conf/repos.toml";
        public const string TransientMetadataPath = "dependencies-metadata.json";
        public const string SourcesPath = "sources";
        public static string SdkPath(string hash) => "sdks/" + hash;
        public static string ArtifactPath(string path) => "dependencies/" + path;

    }
}